README.md
## Introduction

This document presents the Recommendation System, a sophisticated tool designed to optimize data placement in High-Performance Computing (HPC) applications. The system integrates a black box optimization engine, a mathematical model to predict application I/O behavior, a machine learning-based performance model, and a simulation engine. These components work together to explore the parameter space efficiently and simulate application execution within a virtual HPC environment. Our innovative system determines the optimal storage tier for each I/O phase, aiming to reduce the total duration and Burst Buffer usage of workflows. The proof-of-concept implementation has demonstrated promising results, signifying a leap forward for various HPC applications and laying the foundation for further enhancements.

<video controls src="demo/recsys_demo.mp4" title="Title"></video>
---

## System Architecture


*Figure 1: The general architecture of the recommendation system.*

![alt text](demo/RECSYS_D3.2_general_diagram_2.png)
The architecture of the recommendation system is outlined in Figure 12. At the core, the system relies on the Workflow Description File (WDF), which is derived from WP2 and contains intricate details about application steps and storage service requirements. To harness the full potential of the recommendation system, the workflow should have been previously executed with the requisite number of compute nodes, persistent storage services, and with IO Instrumentation activated. The IO Instrumentation database is pivotal for the system as it retains workflow references and job groups, enabling the system to optimize data placement effectively.

---

## System Components

The Recommendation System is composed of several critical components:

- **Application Decomposer**: A tool that analyzes jobs and their I/O behavior from the WDF, constructing a model for each job.

- **Performance Model**: It predicts the latency for each I/O phase across different storage tiers, assessing the impact of storage performance on job I/O.

- **Execution Simulator**: This tool virtually replays the workflow on each storage tier, integrating job representation and anticipated storage performance.

- **Optimization Heuristics**: These heuristics optimize placement combinations and their performance impact, taking into account predicted performance and workflow chronology.

| Component               | Release               | Key Features                                                                                                  |
|-------------------------|-----------------------|---------------------------------------------------------------------------------------------------------------|
| Application Decomposer  | 0.2 (Second Release)  | Analyzes each job’s dataflow timeseries, categorizing I/O and compute phases. Connects to IO Instrumentation.  |
| Performance Model       | 0.1 (First Release)   | Predicts the latency for each I/O phase on each storage tier. Builds a mathematical model of I/O behavior.     |
| Execution Simulator     | 0.2.1 (Second Release)| Virtually replays the workflow on each storage tier, using the job representation and expected storage performance.|
| optimisation Heuristics | 0.1 (First Release)   | Optimises the placement combinations for performance impact, considering workflow chronology and predicted performance. |
| IO Instrumentation Interface | 0.1 (First Release)   | Connects to an IO Instrumentation session to retrieve time-series data for workflow job analysis.                |
| WDF Interface           | 0.0 (Not Released)    | To be developed for integration with Workflow Manager, for structured description and storage service requirements. |


*Table 1: Development status of the components.*

---

## Implementation Process

The Recommendation System's implementation involved identifying necessary components, standardizing data formats, and streamlining data flow between components. The journey from conception to the first version took place over the last 12 months, culminating in the development of key features, testing, and presentation at workshops. As we progress, the focus is now on integrating the remaining elements, including the Workflow Manager from WP2 and the Performability Predictor.

---

## Operational Workflow

The operational workflow of the Recommendation System commences with the App Decomposer analyzing the WDF's job list and extracting their I/O behavior. The Performance Model and Execution Simulator step in to predict I/O latencies and virtually replay the workflow, respectively. The ultimate decisions are made by the Optimization Heuristics, which deduce the most suitable solution for data placement.

---

## Optimization Strategies

The optimization strategies employed by the system are designed to be robust, taking into account the simulation results and a range of possible storage tier placements. The IO Instrumentation Interface serves as a critical link, retrieving job-related data to facilitate optimization.

---

## Main Results

The system has proven effective in reducing workflow durations and optimizing the use of Burst Buffer. The figures accompanying this document illustrate the relative and absolute comparisons of runtime on different tiers against simulation runtime by flavor, underscoring the efficiency of the system.

![alt text](demo/RECSYS_D3.3_LQCD_Pareto_solution.png)

*Figure: This Pareto-front plot demonstrates the trade-off between workflow duration and burst buffer size in a Lattice Quantum Chromodynamics (LQCD) use-case. Each point on the plot represents a unique configuration of job data placement, with the x-axis indicating the total workflow duration and the y-axis showing the burst buffer size required. The color gradient reflects the density of trials: darker areas represent a higher concentration of trial configurations. The red point labeled 'Best Trial' signifies the optimal configuration found that minimizes the workflow duration while using no burst buffer. This optimal solution details the specific job assignments for tier placement and burst buffer usage, marking a significant improvement in performance without increasing resource consumption.*

![alt text](demo/afsm-relative.png)
*Figure: The scatter plot compares the measured runtime of Step BC from an LQCD simulation (in seconds) to the predicted runtime by a recommendation system (RecSys) simulation across different 'flavor' sizes (in GB) on the afsm tier. Each point on the graph represents a comparison of the actual measured runtime (blue dot) against the RecSys simulated runtime (red dot) for a given flavor size. A close clustering of blue and red points suggests a high accuracy of the simulation model in predicting the actual runtime for the LQCD workflow.*


![alt text](demo/p_project-relative.png)
*Figure: This scatter plot presents a comparison of the actual measured runtime of Step BC from an LQCD simulation against the RecSys simulation runtime, categorized by flavor sizes (in GB) on the p_project tier. Blue dots indicate the measured runtime for each flavor, while red dots represent the corresponding simulation runtime predicted by the recommendation system. The plot illustrates the relationship between the computational resource allocation (flavor size) and the efficiency of the simulation in approximating the real-world runtime of LQCD workflows.*

---

## Impact and Benefits

The implementation of the Recommendation System has shown significant impact across various HPC application domains, leading to substantial benefits in workflow efficiency and resource utilization.

---

## Future Directions

Future research aims at furthering the capabilities of the Recommendation System, focusing on improving integration with existing workflow managers, expanding on the Performability Predictor, and advancing the system's overall performance.



jobmodel/CHANGES.md
Change history for Jobmodel
===========================

IOPA 2.0
--------

* Fri Oct 2 2020 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.8

* Fri Jul 31 2020 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.7

* Fri Jul 3 2020 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.6

* Fri Apr 24 2020 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.5

* Fri Mar 27 2020 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.4

* Fri Jan 31 2020 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.3

* Fri Dec 20 2019 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.2
Code refactoring
BRDM-2684: Manage Python library vs product versions

* Fri Nov 29 2019 DataManagement Team <bds-datamanagement@atos.net> 2.0-Bull.1

IOPA 1.2
--------

* Fri Feb 1 2019 DataManagement Team <bds-datamanagement@atos.net> 1.2-Bull.1


cluster_simulator/analytics.py
from cluster_simulator.cluster import Cluster, Tier, bandwidth_share_model, compute_share_model, get_tier, convert_size
from cluster_simulator.phase import DelayPhase, ComputePhase, IOPhase
from cluster_simulator.application import Application
import simpy
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import numpy as np
from itertools import groupby
from operator import itemgetter
from loguru import logger
import itertools


def accumulate_intervals(x_phase, y_phase):
    """Superimpose all points intervall within the boudary"""
    points = sorted(list(set(sorted(list(itertools.chain.from_iterable(x_phase))))))

    y = []
    for point in points:
        # init value for point
        v = 0
        for index, interval in enumerate(x_phase):
            if point >= interval[0] and point < interval[1]:
                v += y_phase[index]
        y.append(v)
    return points, y


def display_apps(data, width=800, height=600):

    DEFAULT_COLORS = ['rgb(31, 119, 180)', 'rgb(255, 127, 14)',
                      'rgb(44, 160, 44)', 'rgb(214, 39, 40)',
                      'rgb(148, 103, 189)', 'rgb(140, 86, 75)',
                      'rgb(227, 119, 194)', 'rgb(127, 127, 127)',
                      'rgb(188, 189, 34)', 'rgb(23, 190, 207)']
    # sort by app key
    items = sorted(data.items, key=itemgetter('app'))
    # list of apps
    apps = list(set(list(data['app'] for data in data.items)))

    fig = make_subplots(rows=len(apps), cols=1, shared_xaxes=True,
                        vertical_spacing=0.2, subplot_titles=apps)
    # iterate on apps
    i = 0
    for app, app_elements in groupby(items,  key=itemgetter('app')):
        # print(f"----------{app}----------")
        # app_color = DEFAULT_COLORS[i]
        offset = 0
        x_app = []
        y_app = []
        text = []

        for phase in app_elements:
            x_app.append(phase["t_start"])
            x_app.append(phase["t_end"])
            y_app.append(phase["bandwidth"])
            y_app.append(phase["bandwidth"])
            placement = "|to/from:"+phase["data_placement"]["placement"] if phase["data_placement"] else ''

            text.append(phase["type"].upper() + placement + "|volume="+convert_size(phase["volume"]))
            text.append(phase["type"].upper() + placement + "|volume="+convert_size(phase["volume"]))

        fig.append_trace(go.Scatter(x=np.array(x_app), y=np.array(y_app),
                                    text=text,
                                    textposition="top center",
                                    name=app, line_shape='hv', showlegend=False), row=i+1, col=1)
        i += 1
        fig.update_xaxes(title_text="time in s")
        fig.update_yaxes(title_text="bandwidth in MB/s")
        fig.update_layout(width=width, height=height, title_text=f"Stacked Volume Time Series for apps:{apps}")
    return fig


def accumulate_intervals(x_phase, y_phase):
    points = sorted(list(set(sorted(list(itertools.chain.from_iterable(x_phase))))))

    y = []
    for point in points:
        # init value for point
        v = 0
        for index, interval in enumerate(x_phase):
            # if intereset point is in interval
            if point >= interval[0] and point < interval[1]:
                # cumulate value from interval having intersections
                v += y_phase[index]
        y.append(v)
    return points, y


def display_cluster(data, cluster, width=800, height=600):
    fig = make_subplots(rows=3, cols=1, shared_xaxes=True,
                        vertical_spacing=0.1)
    # iterate on phases
    i = 0
    x_phase = []
    y_phase = []
    x_tiers = []
    capacities = dict()
    text = []
    cpu_phase = []
    storage = dict()
    buffer_storage = dict()
    for phase in data.items:
        # | Monitoring| app: P6 | type: movement | cpu_usage: 1 | t_start: 281.0 | t_end: 287.5 | bandwidth_concurrency: 1 | bandwidth: 40.0 MB/s | phase_duration: 6.5 | volume: 260.0 MB | tiers: ['HDD'] | data_placement: {'placement': 'HDD', 'source': 'BB'} | init_level: {'HDD': 11740000000.0, 'BB': 8490000000.0} | tier_level: {'HDD': 12000000000.0, 'BB': 8490000000.0} | BB_level: 8490000000.0

        x_phase.append([phase["t_start"], phase["t_end"]])
        y_phase.append(phase["cpu_usage"])
        x_tiers.append(phase["t_start"])
        x_tiers.append(phase["t_end"])

        tier_indication = "|" + phase["data_placement"]["placement"] if phase["data_placement"] else ''

        text.append(phase["app"] + "|" + phase["type"].capitalize() + tier_indication + " | +volume="+convert_size(phase["volume"]))
        text.append(phase["app"] + "|" + phase["type"].capitalize() + tier_indication + " | +volume="+convert_size(phase["volume"]))

        for tier in phase["tiers"]:
            storage.setdefault(tier, []).append(phase["tier_level"][tier])
            storage.setdefault(tier, []).append(phase["tier_level"][tier])

    # trace CPU usage
    points, values = accumulate_intervals(x_phase, y_phase)
    fig.append_trace(go.Scatter(x=np.array(points), y=np.array(values),
                                line_shape='hv', showlegend=False), row=1, col=1)
    fig['layout']['yaxis']['title'] = 'CPU usage'

    for tier in storage.keys():
        i += 1
        # retrieve capacities
        tier_capacity = [cluster.tiers[j].capacity.capacity for j, ctier in enumerate(cluster.tiers) if ctier.name == tier][0]
        print(np.array(x_tiers))
        print(np.array(storage[tier])/tier_capacity)
        fig.append_trace(go.Scatter(x=np.array(x_tiers),
                                    y=100*np.array(storage[tier])/tier_capacity,
                                    text=text,
                                    textposition="top center",
                                    name=tier, line_shape='linear', showlegend=False), row=i+1, col=1)
        fig['layout']['yaxis' + str(i+1)]['title'] = tier + ' usage in %'

        # fig.append_trace(go.Scatter(x=np.array([x_tiers[0], x_tiers[-1]]),
        #                             y=np.array([cluster_tier.capacity.capacity]*2),
        #                             text=["Maximum Tier Capacity=" + str(cluster_tier.capacity.capacity)]*2, line_shape='hv', showlegend=False, line=dict(color='red', width=3, dash='dot')), row=i+1, col=1)

    # for tier in storage.keys():
    #     i += 1
    #     for index, abscissa in enumerate(x_phase):
    #         fig.append_trace(go.Scatter(x=np.array(x_phase[index]),
    #                                     y=np.array(storage[tier][index]),
    #                                     line_shape='linear', showlegend=False),
    #                                     row=1+i, col=1)

    # fig.append_trace(go.Scatter(x=np.array(points), y=np.array([cluster.compute_cores]),
    #                              line_shape='hv', showlegend=False), row=1, col=1)
    fig.append_trace(go.Scatter(x=np.array([points[0], points[-1]]), y=np.array([cluster.compute_cores.capacity]*2), text=["Maximum available cores in cluster=" +
                                                                                                                           str(cluster.compute_cores.capacity)]*2, line_shape='hv', showlegend=False, line=dict(color='red', width=3, dash='dot')), row=1, col=1)

    # fig.update_xaxes(title_text="time in s")
    # fig.update_yaxes(title_text="Units of used cores")

    fig.update_layout(height=height, width=width, title_text="State of the Cluster")
    return fig


def display_run(data, cluster, width=800, height=600):
    # | Monitoring| app: P6 | type: movement | cpu_usage: 1 | t_start: 281.0 | t_end: 287.5 | bandwidth_concurrency: 1 | bandwidth: 40.0 MB/s | phase_duration: 6.5 | volume: 260.0 MB | tiers: ['HDD'] | data_placement: {'placement': 'HDD', 'source': 'BB'} | init_level: {'HDD': 11740000000.0, 'BB': 8490000000.0} | tier_level: {'HDD': 12000000000.0, 'BB': 8490000000.0} | BB_level: 8490000000.0
    # | Monitoring| app: N6 | type: eviction | cpu_usage: 1 | t_start: 93.5 | t_end: 93.5 | bandwidth_concurrency: 3 | bandwidth: inf MB/s | phase_duration: 0 | volume: 740.0 MB | tiers: ['HDD'] | data_placement: {'placement': 'BB'} | init_level: {'HDD': 2740000000.0, 'BB': 10000000000.0} | tier_level: {'HDD': 2740000000.0, 'BB': 9260000000.0} | BB_level: 9260000000.0

    apps = list(set(list(data['app'] for data in data.items)))
    # sort by app key
    items = sorted(data.items, key=itemgetter('app'))
    # list of apps
    list_apps = [app for app, _ in groupby(items,  key=itemgetter('app'))]
    # list Burst Buffers
    bb_tier = [cluster.ephemeral_tier.name + " ("+convert_size(cluster.ephemeral_tier.capacity.capacity)+")"] if cluster.ephemeral_tier else []

    subplots_list = list_apps + bb_tier + [tier.name + " ("+convert_size(tier.capacity.capacity)+")" for tier in cluster.tiers] + ["CPU Cores"]

    fig = make_subplots(rows=len(subplots_list), cols=1, shared_xaxes=True,
                        # vertical_spacing=0.2,
                        subplot_titles=subplots_list)

    # iterate on apps to plot their dataflows
    i = 1
    for app, app_elements in groupby(items,  key=itemgetter('app')):
        # print(f"----------{app}----------")
        # app_color = DEFAULT_COLORS[i]
        offset = 0

        x_app = []
        y_app = []
        text = []

        x_mvt = []
        y_mvt = []
        text_mvt = []

        for phase in app_elements:
            if phase['type'] not in ['movement']:
                x_app.append(phase["t_start"])
                x_app.append(phase["t_end"])
                y_app.append(phase["bandwidth"])
                y_app.append(phase["bandwidth"])
                placement = "|to/from:"+phase["data_placement"]["placement"] if phase["data_placement"] else ''

                text.append(phase["type"].upper() + placement + "|volume="+convert_size(phase["volume"]))
                text.append(phase["type"].upper() + placement + "|volume="+convert_size(phase["volume"]))

            if phase["type"] in ['movement']:
                x_mvt.append(phase["t_start"])
                x_mvt.append(phase["t_end"])
                y_mvt.append(phase["bandwidth"])
                y_mvt.append(phase["bandwidth"])
                placement = "|in:"+phase["data_placement"]["placement"] if phase["data_placement"] else ''
                source = ""
                if "data_placement" in phase and "source" in phase["data_placement"]:
                    source = "|from:"+phase["data_placement"]["source"]

                text_mvt.append(phase["type"].upper() + placement + source + "|volume="+convert_size(phase["volume"]))
                text_mvt.append(phase["type"].upper() + placement + "|volume="+convert_size(phase["volume"]))

                # plot the app phases
        fig.append_trace(go.Scatter(x=np.array(x_app), y=np.array(y_app),
                                    text=text,
                                    textposition="top center",
                                    name="app#"+app, line_shape='linear'), row=i, col=1)
        fig.append_trace(go.Scatter(x=np.array(x_mvt), y=np.array(y_mvt),
                                    text=text_mvt,
                                    textposition="top center",
                                    name=app + " mvt", line_shape='linear', line={'dash': 'dot'}), row=i, col=1)
        fig['layout']['yaxis'+str(i)]['title'] = 'dataflow in MB/s'
        fig.update_xaxes(title_text="time in s")
        i += 1

    x_phase = []
    y_phase = []
    x_tiers = [0]
    x_evict = []
    capacities = dict()
    text = []
    text_evict = []
    cpu_phase = []
    buffer_storage = dict()
    storage = dict()
    bb_levels = dict()
    storage_levels = dict()
    eviction_levels = dict()
    for phase in data.items:

        x_phase.append([phase["t_start"], phase["t_end"]])
        y_phase.append(phase["cpu_usage"])
        x_tiers.append(phase["t_start"])
        x_tiers.append(phase["t_end"])

        # feeding eviction dict
        if phase["type"] == "eviction":
            if "source" in phase["data_placement"]:
                tier = phase["data_placement"]["source"]
                eviction_levels.setdefault(tier, []).append([phase["init_level"][tier], phase["tier_level"][tier]])
                x_evict.append([phase["t_start"], phase["t_end"]])
                text_evict.append(phase["type"].upper() + "|IN:"+tier + "|volume="+convert_size(phase["init_level"][tier]-phase["tier_level"][tier]))
                text_evict.append(phase["type"].upper() + "|IN:"+tier + "|volume="+convert_size(phase["init_level"][tier]-phase["tier_level"][tier]))

        tier_indication = "|" + phase["data_placement"]["placement"] if phase["data_placement"] else ''

        text.append(phase["app"] + "|" + phase["type"].capitalize() + tier_indication + " | +volume="+convert_size(phase["volume"]))

        # feeding tiers level
        for tier in phase["tiers"]:
            if tier not in storage_levels.keys():
                storage_levels[tier] = dict()
            storage_levels[tier][phase["t_start"]] = phase["init_level"][tier]
            storage_levels[tier][phase["t_end"]] = phase["tier_level"][tier]
            storage.setdefault(tier, []).append(phase["init_level"][tier])
            storage.setdefault(tier, []).append(phase["tier_level"][tier])

        # feeding buffer level
        if cluster.ephemeral_tier:
            bb_tier = cluster.ephemeral_tier.name
            if bb_tier+"_level" in phase:
                bb_levels[phase["t_start"]] = phase["init_level"][bb_tier]
                bb_levels[phase["t_end"]] = phase["tier_level"][bb_tier]
                buffer_storage.setdefault(bb_tier, []).append([phase["init_level"][bb_tier], phase["tier_level"][bb_tier]])

    # Burst Buffer level tracing
    for bb in buffer_storage.keys():
        bb_capacity = cluster.ephemeral_tier.capacity.capacity
        # for segment, level in zip(x_phase, buffer_storage[bb]):
        #     fig.append_trace(go.Scatter(x=np.array(segment),
        #                                 y=level,
        #                                 text=text,
        #                                 textposition="top center",
        #                                 name=bb, line_shape='linear', showlegend=False), row=i, col=1)
        fig.append_trace(go.Scatter(x=np.array(list(bb_levels.keys())),
                                    y=100*bb_capacity*np.array(list(bb_levels.values()))/bb_capacity/100,
                                    text=text,
                                    textposition="top center",
                                    name=bb, line_shape='linear', showlegend=False), row=i, col=1)

        if bb in eviction_levels:
            for segment, level in zip(x_evict, eviction_levels[bb]):
                # fig.append_trace(go.Scatter(x=np.array(segment),
                #                             y=level,
                #                             text=text,
                #                             textposition="top center",
                #                             name=bb, line=dict(color='red', width=3, dash='dot'), showlegend=False), row=i, col=1)
                fig.append_trace(go.Scatter(x=np.array(segment), y=np.array(level),
                                            text=text_evict,
                                            textposition="top center",
                                            name=" Eviction",
                                            line=dict(shape='linear', color="black"),
                                            showlegend=False), row=i, col=1)
                # fig.add_shape(type="line", x0=segment[0], y0=level[0], x1=segment[1], y1=level[1],
                #               line=dict(width=3, color="black"),
                #               row=i, col=1)
                # fig.add_annotation(ax=segment[0],
                #                    ay=level[0],
                #                    x=segment[1],
                #                    y=level[1],
                #                    xref='x',

                #                    yref='y',
                #                    axref='x',
                #                    ayref='y',
                #                    text='',  # if you want only the arrow
                #                    showarrow=True,
                #                    arrowhead=3,
                #                    arrowsize=1,
                #                    arrowwidth=1,
                #                    arrowcolor='black', row=i, col=1)

        # fig.append_trace(go.Scatter(x=np.array(list(bb_levels.keys())),
        #                             y=100*bb_capacity*np.array(list(bb_levels.values()))/bb_capacity/100,
        #                             text=text,
        #                             textposition="top center",
        #                             name=bb, line_shape='linear', showlegend=False), row=i, col=1)
        # fig.append_trace(go.Scatter(x=np.array(x_tiers),
        #                             y=100*np.array(buffer_storage[bb])/bb_capacity,
        #                             text=text,
        #                             textposition="top center",
        #                             name=bb, line_shape='linear', showlegend=False), row=i, col=1)
        fig['layout']['yaxis' + str(i)]['title'] = bb + ' usage in Bytes'
        i += 1

    for tier in storage.keys():
        # retrieve capacities
        tier_capacity = [cluster.tiers[j].capacity.capacity for j, ctier in enumerate(cluster.tiers) if ctier.name == tier][0]
        fig.append_trace(go.Scatter(x=np.array(list(storage_levels[tier].keys())),
                                    y=100*tier_capacity*np.array(list(storage_levels[tier].values()))/tier_capacity/100,
                                    text=text,
                                    textposition="top center",
                                    name=tier, line_shape='linear', showlegend=False), row=i, col=1)
        # fig.append_trace(go.Scatter(x=np.array(x_tiers),
        #                             y=100*np.array(storage[tier])/tier_capacity,
        #                             text=text,
        #                             textposition="top center",
        #                             name=tier, line_shape='linear', showlegend=False), row=i, col=1)
        fig['layout']['yaxis' + str(i)]['title'] = tier + ' usage in Bytes'

        i += 1

    # CPU tracing
    points, values = accumulate_intervals(x_phase, y_phase)
    fig.append_trace(go.Scatter(x=np.array(points), y=np.array(values),
                                line_shape='hv', showlegend=False), row=i, col=1)

    fig.append_trace(go.Scatter(x=np.array([points[0], points[-1]]), y=np.array([cluster.compute_cores.capacity]*2), text=["Maximum available cores in cluster=" + str(cluster.compute_cores.capacity)]*2, line_shape='hv', showlegend=False,
                                line=dict(color='red', width=3, dash='dot')), row=i, col=1)
    fig['layout']['yaxis'+str(i)]['title'] = 'CPU usage'
    i += 1

    fig.update_layout(width=width, height=height, title_text="State of the Cluster")

    return fig
